File: typescript/src/grids/cubed_sphere.ts
/**
 * Cubed-sphere grid (gnomonic, 6 panels × Nc × Nc) — accessor runtime.
 *
 * Per `docs/GRIDS_API.md` §2.6 and the 2026-04-20 mayor correction in bead
 * dsc-suu: `toESM()` emits a small declarative config, not a serialized
 * geometry blob. Geometry is derived on demand via the accessors
 * (`cell_centers`, `neighbors`, `metric_eval`, `cell_area`).
 *
 * Panels are 0-indexed internally (0..5) mapping to the standard FV3
 * numbering: 0=front, 1=right, 2=top (north pole), 3=back, 4=left,
 * 5=bottom (south pole). Cell indices i,j are 0-indexed in [0, Nc).
 */

import type { Dtype, Grid } from "./types.js";

export interface CubedSphereOpts {
  Nc: number;
  R?: number;
  dtype?: Dtype;
  ghosts?: number;
}

export type Edge = "west" | "east" | "south" | "north";

export interface NeighborRef {
  panel: number;
  i: number;
  j: number;
  edge: Edge;
}

export type MetricName =
  | "J"
  | "g_xixi"
  | "g_etaeta"
  | "g_xieta"
  | "ginv_xixi"
  | "ginv_etaeta"
  | "ginv_xieta";

export interface CubedSphereGrid extends Grid {
  readonly family: "cubed_sphere";
  readonly topology: "block_structured";
  readonly generator: "gnomonic_c6";
  readonly Nc: number;
  readonly R: number;
  readonly ghosts: number;
  readonly n_panels: 6;
  readonly n_cells: number;

  cell_centers(panel: number, i: number, j: number): { lon: number; lat: number };
  neighbors(panel: number, i: number, j: number): NeighborRef[];
  metric_eval(name: MetricName, panel: number, i: number, j: number): number;
  cell_area(panel: number, i: number, j: number): number;
}

interface PanelNeighbor {
  panel: number;
  edge: Edge;
  reverseIndex: boolean;
}

type ConnectivityRow = Record<Edge, PanelNeighbor>;

// Panel connectivity — mirrors src/grids/panel_connectivity.jl (0-indexed).
// Julia panels 1..6 → TS panels 0..5.
const PANEL_CONNECTIVITY: ReadonlyArray<ConnectivityRow> = [
  // 0 (front)
  {
    west: { panel: 4, edge: "east", reverseIndex: false },
    east: { panel: 1, edge: "west", reverseIndex: false },
    south: { panel: 5, edge: "north", reverseIndex: false },
    north: { panel: 2, edge: "south", reverseIndex: false },
  },
  // 1 (right)
  {
    west: { panel: 0, edge: "east", reverseIndex: false },
    east: { panel: 3, edge: "west", reverseIndex: false },
    south: { panel: 5, edge: "east", reverseIndex: true },
    north: { panel: 2, edge: "east", reverseIndex: false },
  },
  // 2 (top / north pole)
  {
    west: { panel: 4, edge: "north", reverseIndex: true },
    east: { panel: 1, edge: "north", reverseIndex: false },
    south: { panel: 0, edge: "north", reverseIndex: false },
    north: { panel: 3, edge: "north", reverseIndex: true },
  },
  // 3 (back)
  {
    west: { panel: 1, edge: "east", reverseIndex: false },
    east: { panel: 4, edge: "west", reverseIndex: false },
    south: { panel: 5, edge: "south", reverseIndex: true },
    north: { panel: 2, edge: "north", reverseIndex: true },
  },
  // 4 (left)
  {
    west: { panel: 3, edge: "east", reverseIndex: false },
    east: { panel: 0, edge: "west", reverseIndex: false },
    south: { panel: 5, edge: "west", reverseIndex: false },
    north: { panel: 2, edge: "west", reverseIndex: true },
  },
  // 5 (bottom / south pole)
  {
    west: { panel: 4, edge: "south", reverseIndex: false },
    east: { panel: 1, edge: "south", reverseIndex: true },
    south: { panel: 3, edge: "south", reverseIndex: true },
    north: { panel: 0, edge: "south", reverseIndex: false },
  },
];

function gnomonicToCart(
  xi: number,
  eta: number,
  panel: number,
): [number, number, number] {
  const X = Math.tan(xi);
  const Y = Math.tan(eta);
  const D = Math.sqrt(1 + X * X + Y * Y);
  switch (panel) {
    case 0:
      return [1 / D, X / D, Y / D];
    case 1:
      return [-X / D, 1 / D, Y / D];
    case 2:
      return [-Y / D, X / D, 1 / D];
    case 3:
      return [-1 / D, -X / D, Y / D];
    case 4:
      return [X / D, -1 / D, Y / D];
    case 5:
      return [Y / D, X / D, -1 / D];
    default:
      throw new RangeError(`cubed_sphere: invalid panel ${panel} (expected 0..5)`);
  }
}

function gnomonicToLonLat(
  xi: number,
  eta: number,
  panel: number,
): { lon: number; lat: number } {
  const [x, y, z] = gnomonicToCart(xi, eta, panel);
  const clamped = z < -1 ? -1 : z > 1 ? 1 : z;
  return { lon: Math.atan2(y, x), lat: Math.asin(clamped) };
}

function gnomonicMetric(
  xi: number,
  eta: number,
  R: number,
): { J: number; g_xixi: number; g_etaeta: number; g_xieta: number } {
  const X = Math.tan(xi);
  const Y = Math.tan(eta);
  const secXi2 = 1 + X * X;
  const secEta2 = 1 + Y * Y;
  const D2 = 1 + X * X + Y * Y;
  const D2_32 = Math.pow(D2, 1.5);
  const R2 = R * R;
  return {
    J: (R2 * secXi2 * secEta2) / D2_32,
    g_xixi: (R2 * secXi2 * secXi2 * secEta2) / (D2 * D2),
    g_etaeta: (R2 * secEta2 * secEta2 * secXi2) / (D2 * D2),
    g_xieta: (-R2 * X * Y * secXi2 * secEta2) / (D2 * D2),
  };
}

function cellAreaFromCorners(
  corners: ReadonlyArray<[number, number, number]>,
  R: number,
): number {
  // Spherical-excess via inner-angle sum at 4 corners on the unit sphere.
  const n = corners.length;
  let angleSum = 0;
  for (let k = 0; k < n; k++) {
    const vPrev = corners[(k - 1 + n) % n];
    const vCurr = corners[k];
    const vNext = corners[(k + 1) % n];
    const t1 = sideTangent(vCurr, vPrev);
    const t2 = sideTangent(vCurr, vNext);
    const n1 = norm(t1);
    const n2 = norm(t2);
    if (n1 < 1e-15 || n2 < 1e-15) {
      angleSum += Math.PI / 2;
    } else {
      const c = dot(t1, t2) / (n1 * n2);
      angleSum += Math.acos(c < -1 ? -1 : c > 1 ? 1 : c);
    }
  }
  return R * R * (angleSum - (n - 2) * Math.PI);
}

type Vec3 = [number, number, number];

function dot(a: Vec3, b: Vec3): number {
  return a[0] * b[0] + a[1] * b[1] + a[2] * b[2];
}

function cross(a: Vec3, b: Vec3): Vec3 {
  return [
    a[1] * b[2] - a[2] * b[1],
    a[2] * b[0] - a[0] * b[2],
    a[0] * b[1] - a[1] * b[0],
  ];
}

function sub(a: Vec3, b: Vec3): Vec3 {
  return [a[0] - b[0], a[1] - b[1], a[2] - b[2]];
}

function norm(a: Vec3): number {
  return Math.sqrt(dot(a, a));
}

// t = vCurr × ((vOther - vCurr) × vCurr): tangent at vCurr along the
// great-circle arc toward vOther (matches Julia compute_cell_area).
function sideTangent(vCurr: Vec3, vOther: Vec3): Vec3 {
  return cross(vCurr, cross(sub(vOther, vCurr), vCurr));
}

function transformGhostIndex(
  nb: PanelNeighbor,
  i: number,
  j: number,
  Nc: number,
  localEdge: Edge,
): { i: number; j: number } {
  // i,j are 0-indexed in the *local* panel; returns 0-indexed indices in
  // the neighbor panel for the ghost cell just across `localEdge`.
  //
  // "along" = coordinate that varies along the shared edge (0..Nc-1);
  // "perp" = depth into the neighbor (0 = first row off the shared edge).
  let along: number;
  const perp = 0;
  if (localEdge === "west" || localEdge === "east") {
    along = j;
  } else {
    along = i;
  }
  if (nb.reverseIndex) {
    along = Nc - 1 - along;
  }
  switch (nb.edge) {
    case "west":
      return { i: perp, j: along };
    case "east":
      return { i: Nc - 1 - perp, j: along };
    case "south":
      return { i: along, j: perp };
    case "north":
      return { i: along, j: Nc - 1 - perp };
  }
}

function checkPanel(panel: number): void {
  if (!Number.isInteger(panel) || panel < 0 || panel > 5) {
    throw new RangeError(`cubed_sphere: invalid panel ${panel} (expected integer 0..5)`);
  }
}

function checkCell(i: number, j: number, Nc: number): void {
  if (!Number.isInteger(i) || i < 0 || i >= Nc) {
    throw new RangeError(`cubed_sphere: invalid i ${i} (expected integer 0..${Nc - 1})`);
  }
  if (!Number.isInteger(j) || j < 0 || j >= Nc) {
    throw new RangeError(`cubed_sphere: invalid j ${j} (expected integer 0..${Nc - 1})`);
  }
}

export function cubed_sphere(opts: CubedSphereOpts): CubedSphereGrid {
  if (opts === undefined || opts === null) {
    throw new TypeError("cubed_sphere: options object is required");
  }
  const { Nc } = opts;
  if (Nc === undefined) {
    throw new TypeError("cubed_sphere: missing required option 'Nc'");
  }
  if (!Number.isInteger(Nc) || Nc <= 0) {
    throw new RangeError(`cubed_sphere: 'Nc' must be a positive integer (got ${Nc})`);
  }
  const R = opts.R ?? 6.371e6;
  if (!Number.isFinite(R) || R <= 0) {
    throw new RangeError(`cubed_sphere: 'R' must be a positive finite number (got ${R})`);
  }
  const dtype: Dtype = opts.dtype ?? "float64";
  if (dtype !== "float64" && dtype !== "float32") {
    throw new RangeError(`cubed_sphere: 'dtype' must be 'float64' or 'float32' (got ${String(dtype)})`);
  }
  const ghosts = opts.ghosts ?? 0;
  if (!Number.isInteger(ghosts) || ghosts < 0) {
    throw new RangeError(`cubed_sphere: 'ghosts' must be a non-negative integer (got ${ghosts})`);
  }

  const dxi = Math.PI / 2 / Nc;
  const xiCenter = (i: number): number => -Math.PI / 4 + (i + 0.5) * dxi;
  const xiEdge = (i: number): number => -Math.PI / 4 + i * dxi;

  const grid: CubedSphereGrid = {
    family: "cubed_sphere",
    topology: "block_structured",
    generator: "gnomonic_c6",
    Nc,
    R,
    dtype,
    ghosts,
    n_panels: 6,
    n_cells: 6 * Nc * Nc,

    cell_centers(panel: number, i: number, j: number) {
      checkPanel(panel);
      checkCell(i, j, Nc);
      return gnomonicToLonLat(xiCenter(i), xiCenter(j), panel);
    },

    neighbors(panel: number, i: number, j: number): NeighborRef[] {
      checkPanel(panel);
      checkCell(i, j, Nc);
      const refs: NeighborRef[] = [];
      const conn = PANEL_CONNECTIVITY[panel];
      // west
      if (i > 0) {
        refs.push({ panel, i: i - 1, j, edge: "west" });
      } else {
        const t = transformGhostIndex(conn.west, i, j, Nc, "west");
        refs.push({ panel: conn.west.panel, i: t.i, j: t.j, edge: "west" });
      }
      // east
      if (i < Nc - 1) {
        refs.push({ panel, i: i + 1, j, edge: "east" });
      } else {
        const t = transformGhostIndex(conn.east, i, j, Nc, "east");
        refs.push({ panel: conn.east.panel, i: t.i, j: t.j, edge: "east" });
      }
      // south
      if (j > 0) {
        refs.push({ panel, i, j: j - 1, edge: "south" });
      } else {
        const t = transformGhostIndex(conn.south, i, j, Nc, "south");
        refs.push({ panel: conn.south.panel, i: t.i, j: t.j, edge: "south" });
      }
      // north
      if (j < Nc - 1) {
        refs.push({ panel, i, j: j + 1, edge: "north" });
      } else {
        const t = transformGhostIndex(conn.north, i, j, Nc, "north");
        refs.push({ panel: conn.north.panel, i: t.i, j: t.j, edge: "north" });
      }
      return refs;
    },

    metric_eval(name: MetricName, panel: number, i: number, j: number): number {
      checkPanel(panel);
      checkCell(i, j, Nc);
      const m = gnomonicMetric(xiCenter(i), xiCenter(j), R);
      switch (name) {
        case "J":
          return m.J;
        case "g_xixi":
          return m.g_xixi;
        case "g_etaeta":
          return m.g_etaeta;
        case "g_xieta":
          return m.g_xieta;
        case "ginv_xixi":
        case "ginv_etaeta":
        case "ginv_xieta": {
          const det = m.g_xixi * m.g_etaeta - m.g_xieta * m.g_xieta;
          if (name === "ginv_xixi") return m.g_etaeta / det;
          if (name === "ginv_etaeta") return m.g_xixi / det;
          return -m.g_xieta / det;
        }
        default:
          throw new RangeError(`cubed_sphere: unknown metric '${String(name)}'`);
      }
    },

    cell_area(panel: number, i: number, j: number): number {
      checkPanel(panel);
      checkCell(i, j, Nc);
      const xw = xiEdge(i);
      const xe = xiEdge(i + 1);
      const ys = xiEdge(j);
      const yn = xiEdge(j + 1);
      const corners: Vec3[] = [
        gnomonicToCart(xw, ys, panel),
        gnomonicToCart(xe, ys, panel),
        gnomonicToCart(xe, yn, panel),
        gnomonicToCart(xw, yn, panel),
      ];
      return cellAreaFromCorners(corners, R);
    },

    toESM(): object {
      return {
        family: "cubed_sphere",
        version: "1.0.0",
        dtype,
        topology: "block_structured",
        generator: "gnomonic_c6",
        options: { Nc, R, ghosts },
        n_panels: 6,
        n_cells: 6 * Nc * Nc,
      };
    },
  };

  return grid;
}

File: python/src/earthsci_discretizations/grids/cubed_sphere.py
"""Cubed-sphere grid accessor runtime (FV3 gnomonic convention).

Conforms to the cross-binding contract in ``docs/GRIDS_API.md`` §2.4, §3.2, §7.

Per the 2026-04-20 scope correction, the ``.esm`` lowering is a small
declarative config (family, dimensions, extents, generator reference), not a
serialized geometry blob. Geometry is derived on demand by the accessors:

* :meth:`CubedSphereGrid.cell_centers` - cell-center (lon, lat)
* :meth:`CubedSphereGrid.neighbors` - 4-way face connectivity, including
  cross-panel edges
* :meth:`CubedSphereGrid.metric_eval` - metric-tensor components at cell
  centers (covariant, inverse, and derived quantities like face area)

Cross-binding conformance (``docs/GRIDS_API.md`` §4) compares these accessor
outputs at pinned query points rather than serialized bytes.
"""

from __future__ import annotations

import math
from collections.abc import Mapping

import numpy as np

__all__ = ["cubed_sphere", "CubedSphereGrid"]


_DTYPE_MAP: Mapping[str, type] = {"float64": np.float64, "float32": np.float32}

_DIRECTIONS = ("W", "E", "S", "N")

# Standard cubed-sphere panel connectivity (0-indexed).
# Mirrors src/grids/panel_connectivity.jl (Julia's 1..6 shifted to Python's 0..5).
# Entry: (neighbor_panel, neighbor_edge, reverse_along_edge)
_PANEL_CONNECTIVITY: Mapping[int, Mapping[str, tuple[int, str, bool]]] = {
    0: {"W": (4, "E", False), "E": (1, "W", False),
        "S": (5, "N", False), "N": (2, "S", False)},
    1: {"W": (0, "E", False), "E": (3, "W", False),
        "S": (5, "E", True),  "N": (2, "E", False)},
    2: {"W": (4, "N", True),  "E": (1, "N", False),
        "S": (0, "N", False), "N": (3, "N", True)},
    3: {"W": (1, "E", False), "E": (4, "W", False),
        "S": (5, "S", True),  "N": (2, "N", True)},
    4: {"W": (3, "E", False), "E": (0, "W", False),
        "S": (5, "W", False), "N": (2, "W", True)},
    5: {"W": (4, "S", False), "E": (1, "S", True),
        "S": (3, "S", True),  "N": (0, "S", False)},
}


def _gnomonic_to_cart(xi: float, eta: float, panel: int) -> tuple[float, float, float]:
    """Unit-sphere Cartesian (x, y, z) for computational coords (xi, eta) on ``panel``."""
    X = math.tan(xi)
    Y = math.tan(eta)
    r = math.sqrt(1.0 + X * X + Y * Y)
    if panel == 0:
        return (1.0 / r, X / r, Y / r)
    if panel == 1:
        return (-X / r, 1.0 / r, Y / r)
    if panel == 2:
        return (-Y / r, X / r, 1.0 / r)
    if panel == 3:
        return (-1.0 / r, -X / r, Y / r)
    if panel == 4:
        return (X / r, -1.0 / r, Y / r)
    if panel == 5:
        return (Y / r, X / r, -1.0 / r)
    raise ValueError(f"invalid panel: {panel!r}")


def _gnomonic_to_lonlat(xi: float, eta: float, panel: int) -> tuple[float, float]:
    x, y, z = _gnomonic_to_cart(xi, eta, panel)
    lon = math.atan2(y, x)
    lat = math.asin(max(-1.0, min(1.0, z)))
    return lon, lat


def _gnomonic_metric(xi: float, eta: float, R: float) -> tuple[float, float, float, float]:
    """Covariant metric ``(J, g_xixi, g_etaeta, g_xieta)`` at (xi, eta) for radius R.

    The gnomonic metric is panel-independent: each panel is a rigid rotation of
    the sphere, so the (xi, eta)->sphere metric is the same on all six panels.
    """
    X = math.tan(xi)
    Y = math.tan(eta)
    D2 = 1.0 + X * X + Y * Y
    sx = 1.0 + X * X
    sy = 1.0 + Y * Y
    D3 = D2 * math.sqrt(D2)
    D4 = D2 * D2
    R2 = R * R
    J = R2 * sx * sy / D3
    gxx = R2 * sx * sx * sy / D4
    gyy = R2 * sy * sy * sx / D4
    gxy = -R2 * X * Y * sx * sy / D4
    return J, gxx, gyy, gxy


def _spherical_quad_area(corners: list[tuple[float, float, float]], R: float) -> float:
    """Area of a spherical quadrilateral given 4 unit-sphere corner vectors and radius.

    Sums interior angles via the planar-projection approach in
    ``src/grids/metric_tensors.jl::compute_cell_area`` - stable for the small
    non-planar quads at cube corners.
    """
    n = 4
    total = 0.0
    for k in range(n):
        v_prev = np.asarray(corners[(k - 1) % n], dtype=np.float64)
        v_curr = np.asarray(corners[k], dtype=np.float64)
        v_next = np.asarray(corners[(k + 1) % n], dtype=np.float64)
        t1 = np.cross(v_curr, np.cross(v_prev - v_curr, v_curr))
        t2 = np.cross(v_curr, np.cross(v_next - v_curr, v_curr))
        n1 = float(np.linalg.norm(t1))
        n2 = float(np.linalg.norm(t2))
        if n1 < 1e-15 or n2 < 1e-15:
            total += math.pi / 2
        else:
            cos_ang = float(np.dot(t1, t2)) / (n1 * n2)
            total += math.acos(max(-1.0, min(1.0, cos_ang)))
    return R * R * (total - (n - 2) * math.pi)


_METRIC_NAMES = frozenset({
    "J", "g_xixi", "g_etaeta", "g_xieta",
    "ginv_xixi", "ginv_etaeta", "ginv_xieta",
    "area",
})


class CubedSphereGrid:
    """Gnomonic cubed-sphere grid (6 panels x Nc x Nc cells).

    See ``docs/GRIDS_API.md`` §2.4, §3.2, §7. The grid stores only the
    declarative parameters plus pre-tabulated 1D axis arrays. All geometric
    quantities (lon, lat, metric, area) are derived on demand from the
    gnomonic_c6 generator rather than materialized eagerly; bulk properties
    cache arrays on first access.
    """

    family = "cubed_sphere"
    topology = "block_structured"

    def __init__(
        self,
        *,
        Nc: int,
        R: float,
        dtype: str,
        ghosts: int,
        xi_edges: np.ndarray,
        eta_edges: np.ndarray,
        xi_centers: np.ndarray,
        eta_centers: np.ndarray,
    ) -> None:
        self.Nc = Nc
        self.R = R
        self.dtype = dtype
        self.ghosts = ghosts
        self.xi_edges = xi_edges
        self.eta_edges = eta_edges
        self.xi_centers = xi_centers
        self.eta_centers = eta_centers
        self._lon_cache: np.ndarray | None = None
        self._lat_cache: np.ndarray | None = None
        self._area_cache: np.ndarray | None = None

    @property
    def n_cells(self) -> int:
        return 6 * self.Nc * self.Nc

    @property
    def provenance(self) -> dict:
        import earthsci_discretizations

        return {
            "binding": "python",
            "binding_version": earthsci_discretizations.__version__,
            "source": "earthsci_discretizations.grids.cubed_sphere",
            "generator": "gnomonic_c6",
        }

    def cell_centers(
        self, p: int | None = None, i: int | None = None, j: int | None = None
    ) -> tuple[float, float] | tuple[np.ndarray, np.ndarray]:
        """Cell-center geographic coords.

        With no arguments, returns ``(lon, lat)`` arrays of shape ``(6, Nc, Nc)``.
        With ``(p, i, j)``, returns the scalar ``(lon, lat)`` pair for that cell.
        """
        if p is None and i is None and j is None:
            return self.lon, self.lat
        self._check_cell(p, i, j)
        return _gnomonic_to_lonlat(
            float(self.xi_centers[i]), float(self.eta_centers[j]), int(p)
        )

    def neighbors(self, p: int, i: int, j: int) -> dict[str, tuple[int, int, int]]:
        """Face neighbors of cell ``(p, i, j)`` as ``{dir: (p', i', j')}``.

        Interior cells resolve within the same panel; boundary cells cross to
        the neighbor panel via :data:`_PANEL_CONNECTIVITY`, with the along-edge
        index reversed when the edges are oriented opposite.
        """
        self._check_cell(p, i, j)
        Nc = self.Nc
        out: dict[str, tuple[int, int, int]] = {}
        out["W"] = (p, i - 1, j) if i > 0 else self._cross_panel_neighbor(p, "W", i, j)
        out["E"] = (
            (p, i + 1, j) if i < Nc - 1 else self._cross_panel_neighbor(p, "E", i, j)
        )
        out["S"] = (p, i, j - 1) if j > 0 else self._cross_panel_neighbor(p, "S", i, j)
        out["N"] = (
            (p, i, j + 1) if j < Nc - 1 else self._cross_panel_neighbor(p, "N", i, j)
        )
        return out

    def metric_eval(self, name: str, p: int, i: int, j: int) -> float:
        """Evaluate a metric field at cell center ``(p, i, j)``.

        Valid ``name`` values:

        * ``"J"`` - Jacobian determinant
        * ``"g_xixi"``, ``"g_etaeta"``, ``"g_xieta"`` - covariant metric
        * ``"ginv_xixi"``, ``"ginv_etaeta"``, ``"ginv_xieta"`` - inverse metric
        * ``"area"`` - cell face area (uses panel-specific corner geometry)

        The covariant/inverse metric is panel-independent; ``area`` varies
        only at cube-corner cells where the quadrilateral is non-planar.
        """
        if name not in _METRIC_NAMES:
            raise ValueError(f"unknown metric name: {name!r}")
        self._check_cell(p, i, j)
        if name == "area":
            return self._cell_area(int(p), int(i), int(j))
        xi = float(self.xi_centers[i])
        eta = float(self.eta_centers[j])
        J, gxx, gyy, gxy = _gnomonic_metric(xi, eta, self.R)
        if name == "J":
            return J
        if name == "g_xixi":
            return gxx
        if name == "g_etaeta":
            return gyy
        if name == "g_xieta":
            return gxy
        det = gxx * gyy - gxy * gxy
        if name == "ginv_xixi":
            return gyy / det
        if name == "ginv_etaeta":
            return gxx / det
        return -gxy / det  # ginv_xieta

    @property
    def lon(self) -> np.ndarray:
        if self._lon_cache is None:
            self._materialize_centers()
        assert self._lon_cache is not None
        return self._lon_cache

    @property
    def lat(self) -> np.ndarray:
        if self._lat_cache is None:
            self._materialize_centers()
        assert self._lat_cache is not None
        return self._lat_cache

    @property
    def area(self) -> np.ndarray:
        if self._area_cache is None:
            np_dtype = _DTYPE_MAP[self.dtype]
            arr = np.zeros((6, self.Nc, self.Nc), dtype=np_dtype)
            for p in range(6):
                for i in range(self.Nc):
                    for j in range(self.Nc):
                        arr[p, i, j] = self._cell_area(p, i, j)
            self._area_cache = arr
        return self._area_cache

    def to_esm(self) -> dict:
        """Declarative ``.esm`` lowering per the 2026-04-20 scope correction.

        Returns a §6-schema-shaped config: family, dimensions, extents, and
        a named generator reference. No inline geometry arrays.
        """
        return {
            "family": self.family,
            "version": "1.0.0",
            "dtype": self.dtype,
            "topology": self.topology,
            "generator": "gnomonic_c6",
            "params": {
                "Nc": int(self.Nc),
                "R": float(self.R),
                "ghosts": int(self.ghosts),
            },
            "provenance": self.provenance,
        }

    def __repr__(self) -> str:
        return (
            f"CubedSphereGrid(Nc={self.Nc}, R={self.R}, "
            f"dtype={self.dtype!r}, ghosts={self.ghosts})"
        )

    # internals -----------------------------------------------------------

    def _check_cell(self, p, i, j) -> None:
        if p is None or i is None or j is None:
            raise TypeError("cell index (p, i, j) required")
        if not 0 <= int(p) < 6:
            raise ValueError(f"panel out of range [0, 6): {p!r}")
        if not 0 <= int(i) < self.Nc:
            raise ValueError(f"i out of range [0, {self.Nc}): {i!r}")
        if not 0 <= int(j) < self.Nc:
            raise ValueError(f"j out of range [0, {self.Nc}): {j!r}")

    def _cross_panel_neighbor(
        self, p: int, direction: str, i: int, j: int
    ) -> tuple[int, int, int]:
        nb_panel, nb_edge, reverse = _PANEL_CONNECTIVITY[p][direction]
        a = j if direction in ("W", "E") else i
        if reverse:
            a = self.Nc - 1 - a
        last = self.Nc - 1
        if nb_edge == "W":
            return (nb_panel, 0, a)
        if nb_edge == "E":
            return (nb_panel, last, a)
        if nb_edge == "S":
            return (nb_panel, a, 0)
        return (nb_panel, a, last)

    def _cell_area(self, p: int, i: int, j: int) -> float:
        xi_w = float(self.xi_edges[i])
        xi_e = float(self.xi_edges[i + 1])
        eta_s = float(self.eta_edges[j])
        eta_n = float(self.eta_edges[j + 1])
        corners = [
            _gnomonic_to_cart(xi_w, eta_s, p),
            _gnomonic_to_cart(xi_e, eta_s, p),
            _gnomonic_to_cart(xi_e, eta_n, p),
            _gnomonic_to_cart(xi_w, eta_n, p),
        ]
        return _spherical_quad_area(corners, self.R)

    def _materialize_centers(self) -> None:
        Nc = self.Nc
        np_dtype = _DTYPE_MAP[self.dtype]
        lon = np.zeros((6, Nc, Nc), dtype=np_dtype)
        lat = np.zeros((6, Nc, Nc), dtype=np_dtype)
        xi_c = np.asarray(self.xi_centers, dtype=np.float64)
        eta_c = np.asarray(self.eta_centers, dtype=np.float64)
        Xi, Eta = np.meshgrid(xi_c, eta_c, indexing="ij")
        X = np.tan(Xi)
        Y = np.tan(Eta)
        r = np.sqrt(1.0 + X * X + Y * Y)
        # Per-panel Cartesian (x, y, z) on unit sphere, vectorized.
        panel_xyz = [
            (1.0 / r, X / r, Y / r),
            (-X / r, np.ones_like(r) / r, Y / r),
            (-Y / r, X / r, np.ones_like(r) / r),
            (-np.ones_like(r) / r, -X / r, Y / r),
            (X / r, -np.ones_like(r) / r, Y / r),
            (Y / r, X / r, -np.ones_like(r) / r),
        ]
        for p, (xv, yv, zv) in enumerate(panel_xyz):
            lon[p] = np.arctan2(yv, xv).astype(np_dtype, copy=False)
            lat[p] = np.arcsin(np.clip(zv, -1.0, 1.0)).astype(np_dtype, copy=False)
        self._lon_cache = lon
        self._lat_cache = lat


def cubed_sphere(
    *,
    Nc,
    R: float = 6.371e6,
    dtype="float64",
    ghosts: int = 0,
) -> CubedSphereGrid:
    """Generate a gnomonic cubed-sphere grid (6 panels x Nc x Nc cells).

    See ``docs/GRIDS_API.md`` §2.4 for the cross-binding signature contract.
    All options are keyword-only; ``Nc`` is required. The returned grid
    lowers to a small declarative ``.esm`` via :meth:`CubedSphereGrid.to_esm`;
    geometry is derived on demand.

    Parameters
    ----------
    Nc : int
        Cells per panel edge. Must be >= 1.
    R : float, optional
        Sphere radius in meters. Default 6.371e6 (Earth).
    dtype : {"float64", "float32"} or numpy.dtype, optional
        Element precision. Default "float64".
    ghosts : int, optional
        Halo cell width. Default 0.
    """
    if isinstance(Nc, bool) or not isinstance(Nc, (int, np.integer)):
        raise TypeError(f"Nc must be int, got {type(Nc).__name__}")
    Nc = int(Nc)
    if Nc < 1:
        raise ValueError(f"Nc must be >= 1, got {Nc}")

    if isinstance(ghosts, bool) or not isinstance(ghosts, (int, np.integer)):
        raise TypeError(f"ghosts must be int, got {type(ghosts).__name__}")
    ghosts = int(ghosts)
    if ghosts < 0:
        raise ValueError(f"ghosts must be >= 0, got {ghosts}")

    if isinstance(dtype, np.dtype):
        if dtype == np.float64:
            dtype = "float64"
        elif dtype == np.float32:
            dtype = "float32"
        else:
            raise ValueError(f"unsupported dtype: {dtype}")
    if dtype not in _DTYPE_MAP:
        raise ValueError(f"dtype must be 'float64' or 'float32', got {dtype!r}")

    R = float(R)
    if not (R > 0 and math.isfinite(R)):
        raise ValueError(f"R must be a positive finite number, got {R!r}")

    np_dtype = _DTYPE_MAP[dtype]
    dxi = (math.pi / 2) / Nc
    xi_edges = np.asarray(
        [-math.pi / 4 + k * dxi for k in range(Nc + 1)], dtype=np_dtype
    )
    eta_edges = xi_edges.copy()
    xi_centers = 0.5 * (xi_edges[:-1] + xi_edges[1:])
    eta_centers = xi_centers.copy()

    return CubedSphereGrid(
        Nc=Nc,
        R=R,
        dtype=dtype,
        ghosts=ghosts,
        xi_edges=xi_edges,
        eta_edges=eta_edges,
        xi_centers=xi_centers,
        eta_centers=eta_centers,
    )
